--- opna.cpp	2001/03/30 16:58:58
+++ opna.cpp	2001/07/20 17:40:19
@@ -460,6 +460,8 @@
 	adpcmvol = 0;
 	control2 = 0;
 
+	adpcmmute = false;
+
 	interpolation = false;
 	
 	MakeTable();
@@ -1055,8 +1057,10 @@
 						break;
 				}
 				int s = (adplc * apout0 + (8192-adplc) * apout1) >> 13;
-				StoreSample(dest[0], s & maskl);
-				StoreSample(dest[1], s & maskr);
+				if (!adpcmmute) {
+					StoreSample(dest[0], s & maskl);
+					StoreSample(dest[1], s & maskr);
+				}
 				dest += 2;
 				adplc -= adpld;
 			}
@@ -1068,8 +1072,10 @@
 					adplc += 8192;
 				}
 				int s = (adplc * apout1) >> 13;
-				StoreSample(dest[0], s & maskl);
-				StoreSample(dest[1], s & maskr);
+				if (!adpcmmute) {
+					StoreSample(dest[0], s & maskl);
+					StoreSample(dest[1], s & maskr);
+				}
 				dest += 2;
 				adplc -= adpld;
 			}
@@ -1326,6 +1332,7 @@
 		rhythm[i].pos = 0;
 		rhythm[i].size = 0;
 		rhythm[i].volume = 0;
+		rhythm[i].mute = false;
 	}
 	rhythmtvol = 0;
 	adpcmmask = 0x3ffff;
@@ -1573,8 +1580,10 @@
 				{
 					int sample = (r.sample[r.pos / 1024] * vol) >> 12;
 					r.pos += r.step;
-					StoreSample(dest[0], sample & maskl);
-					StoreSample(dest[1], sample & maskr);
+					if (!r.mute) {
+						StoreSample(dest[0], sample & maskl);
+						StoreSample(dest[1], sample & maskr);
+					}
 				}
 			}
 		}
@@ -1619,6 +1628,14 @@
 	RhythmMix(buffer, nsamples);
 }
 
+void OPNA::SetChannelMask(uint mask)
+{
+	OPNABase::SetChannelMask(mask);
+	adpcmmute = !!(mask & (1<<9));
+	for (int i=0; i<6; i++)
+		rhythm[i].mute = !!(mask & (1<<(10+i)));
+}
+
 #endif // BUILD_OPNA
 
 // ---------------------------------------------------------------------------
@@ -1646,6 +1663,7 @@
 		adpcma[i].stop = 0;
 		adpcma[i].adpcmx = 0;
 		adpcma[i].adpcmd = 0;
+		adpcma[i].mute = false;
 	}
 	adpcmatl = 0;
 	adpcmakey = 0;
@@ -1967,8 +1985,10 @@
 						r.adpcmd = Limit(r.adpcmd, 48*16, 0);
 					}
 					int sample = (r.adpcmx * vol) >> 10;
-					StoreSample(dest[0], sample & maskl);
-					StoreSample(dest[1], sample & maskr);
+					if (!r.mute) {
+						StoreSample(dest[0], sample & maskl);
+						StoreSample(dest[1], sample & maskr);
+					}
 				}
 			}
 		}
@@ -2011,6 +2031,14 @@
 	psg.Mix(buffer, nsamples);
 	ADPCMBMix(buffer, nsamples);
 	ADPCMAMix(buffer, nsamples);
+}
+
+void OPNB::SetChannelMask(uint mask)
+{
+	OPNABase::SetChannelMask(mask);
+	for (int i=0; i<6; i++)
+		adpcma[i].mute = !!(mask & (1<<(9+i)));
+	adpcmmute = !!(mask & (1<<15));
 }
 
 #endif // BUILD_OPNB
